### ISSUE TITLE
Add support for Right to Left text

### ISSUE THREAD

**Issue by andydotxyz** [fyne-io/fyne#90]
We also need to support text that is a combination of LtR and RtL...

See issue #38 for some additional thoughts

**Comment by meyt** [comment]
@andydotxyz There is a lot of work already done about text-reshaping in [HarfBuzz](https://github.com/harfbuzz/harfbuzz).

**Comment by javad-majidi** [comment]
i have made a simple farsi/arabic [package](https://github.com/javad-majidi/farsi-reshaper) that solves this issue

**Comment by andydotxyz** [comment]
There is a lot more involved than just the shaping, which is under way already.
We have to handle the aspects of editing among others.
But I wonder if the comment may have been automated as it was pasted in to 5 issues in quick succession...

**Comment by javad-majidi** [comment]
yes but part of this issue is because it can not be viewed at all that can be handled by FYNE_FONT Env and part of it is because it can not be shown properly (even if you reverse the letters you wouldn’t get correct RTL text because some letters needs to be concatenated to others) that my package resolves that and part of it is for RTL support when for example a sentence is Wrapped by its length it breaks naturally that is not handled by anything so far!

**Comment by andydotxyz** [comment]
I don't disagree with anything you say - but it is a stretch to summarise the whole issue as "solved" by your library.

**Comment by javad-majidi** [comment]
actually i meant solved for showing RTL languages properly rather than solving RTL support and i’ve sent them to various open issues related to “not showing rtl languages properly” to signify that from now on it can at least be shown correctly! that was my bad to copy and paste without regarding the exact issue topic!

**Comment by Jacalz** [comment]
That kind of mass commenting looks a lot like spam in some ways. Please avoid doing so in the future.

**Comment by andydotxyz** [comment]
> actually i meant solved for showing RTL languages properly

It is also worth noting that we cannot depend on, or use code from, a package with no license (or incompatible).
I note also that you extracted the code from a website so I doubt very much that the appropriate copyright/permission is secured that would allow us to use the result.

**Comment by dweymouth** [comment]
Has RTL support landed upstream in go-text? This could be a good ticket to move into the i18n work plan

**Comment by dweymouth** [comment]
I got some context from a Supersonic user on how RTL support should work with mixed-script strings. Here is a screenshot from a web UI showing proper RTL in a mixed-script string:

![image](https://github.com/fyne-io/fyne/assets/5303666/c7257ece-79a1-4484-b7a8-078498b1ce74)

And here is how it currently renders in Supersonic:

![image](https://github.com/fyne-io/fyne/assets/5303666/afb0cc25-17f1-48de-a939-7debcdddfd4c)

So, within a single string, a substring in an RTL language needs to be rendered in proper RTL order, while the rest of the string remains LTR

**Comment by andydotxyz** [comment]
> Has RTL support landed upstream in go-text? This could be a good ticket to move into the i18n work plan

It is landed and also into the go-text/render package. But it requires work in localisation as well which is super early, so might be tight to get it into the same release/

**Comment by dweymouth** [comment]
Agree that it might be tight to get in same release, though part of it doesn't require localization I assume? The mixed-string case I described above - it seems that just detecting a substring that's in an RTL script should always be rendered in RTL order regardless of locale/etc.

**Comment by andydotxyz** [comment]
The always is subjective sadly. The locale info defines the default - so yes it can still matter.

**Comment by mustafa9514** [comment]
Please add support for Right to Left text. Arabic letters cannot be displayed.

**Comment by andydotxyz** [comment]
> Please add support for Right to Left text. Arabic letters cannot be displayed.

I replied to the other instance where you added this request https://github.com/fyne-io/fyne/issues/1623#issuecomment-2125615467.

Note that this issue was open for that request already so I don't know what your message added?

**Comment by nokhodian** [comment]
> i have made a simple farsi/arabic [package](https://github.com/javad-majidi/farsi-reshaper) that solves this issue

that's great work like charm I tried millions of things but this is very smooth thanks a lot
